client/web/src/hooks/useScrollManager.ts

```diff
@@ -0,0 +1,164 @@
+import { useEffect } from 'react'
+
+import { debounce } from 'lodash'
+import { useLocation } from 'react-router'
+
+/**
+ * Object containing maps of saved scroll locations for each path name, which are keyed by the scrollable container they belong to.
+ * This will allow us to easily support scroll preservation in multiple containers, even when they are used in the same page.
+ */
+const SAVED_SCROLL_POSITIONS: { [k: string]: Map<string, number> } = {}
+
+/**
+ * Object containing maps of cached scroll attempts for each path name, which are keyed by the scrollable container they belong to.
+ * This can be used to cancel a scroll attempt if a new attempt comes in for the same path.
+ */
+const CACHED_SCROLL_ATTEMPTS: { [k: string]: Map<string, MutationObserverPromise> } = {}
+
+/** The length of the timeout in milliseconds until we stop trying to scroll. */
+const SCROLL_RETRY_TIMEOUT = 3000
+
+export function useScrollManager(containerKey: string, containerRef: React.RefObject<HTMLElement>): void {
+    const { pathname } = useLocation()
+
+    // Set up the maps for this containerKey if they haven't been created yet
+    useEffect(() => {
+        if (!SAVED_SCROLL_POSITIONS[containerKey]) {
+            SAVED_SCROLL_POSITIONS[containerKey] = new Map<string, number>()
+        }
+
+        if (!CACHED_SCROLL_ATTEMPTS[containerKey]) {
+            CACHED_SCROLL_ATTEMPTS[containerKey] = new Map<string, MutationObserverPromise>()
+        }
+    })
```

Comment:
```suggestion
    }, [containerKey])
```

This should only run on initial mount / if the key changes, right?

Reply:
Pushed this up with my change, feel free to revert if I missed anything :)